### PR TITLE
feature/MCU_module_handleFramesTask

### PR DESCRIPTION
**Pull request by Mihnea9901**
Added method to handle single and multi frames.
Added checks for negative responses.
Added method to handle negative responses.

### CODE REVIEW

**Review by GeorgeCimpoeru — APPROVED** [review]

File: src/mcu/src/HandleFrames.cpp [hunk 1]
```diff
@@ -7,71 +7,450 @@
...
+        if (frame.data[1] == 0x7F)
+        {
+            sid = frame.data[2];
+        }
+        else
+        {
+            sid = frame.data[1];  
+        }
+        std::cout << "Single Frame received:" << std::endl;
+        for (int data_pos = 0; data_pos < frame.can_dlc; ++data_pos)
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:40]: Not uint8_t?

File: src/mcu/src/HandleFrames.cpp [hunk 2]
```diff
@@ -7,71 +7,450 @@
...
+        expected_payload = frame.data[1];
+        /* Number of expected frames */
+        uint8_t expected_frames = frame.data[1] / 7;
+        if(frame.data[1] % 7 > 0)
+        {
+            expected_frames++;
+        }
+        /* get SID from the first frame */
+        sid = frame.data[2];
+        std::cout << "Multi-frame Sequence with " << (int)expected_frames << " frames:" << std::endl;
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:64]: same here

File: src/mcu/src/HandleFrames.cpp [hunk 3]
```diff
@@ -7,71 +7,450 @@
...
+
+        /* Clear the multi_frame_data vector when receiving the first frame */
+        frame_data.clear();
+
+        /* Concatenate the data from the first frame into multi_frame_data */
+        for (int data_pos = 5; data_pos < frame.can_dlc; ++data_pos)
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:70]: also here

File: src/mcu/src/HandleFrames.cpp [hunk 4]
```diff
@@ -7,71 +7,450 @@
...
+        {
+            /* Ignore consecutive frames until the first frame is received */
+            return;
+        }
+        /* Check if the frame's sequence number matches the expected sequence number */
+        if (frame.data[0] != expected_sequence_number) 
+        {
+            std::cerr << "Invalid consecutive frame sequence: expected " 
+                      << std::hex << int(expected_sequence_number) << " but received " 
+                      << std::hex << int(frame.data[0]) << std::endl;
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:91]: and the 2 above?

File: src/mcu/src/HandleFrames.cpp [hunk 5]
```diff
@@ -7,71 +7,450 @@
...
+        {
+            frame_data.push_back(frame.data[data_pos]);
+        }
+        /* Increment sequenceNumber after each concatenation*/
+        expected_sequence_number++;
+         /* Check if all multi-frames have been received */
+        if (frame_data.size() >= expected_payload) 
+        { 
+            std::cout << "data is: ";
+            for (int data_pos = 0; data_pos < (int)frame_data.size(); ++data_pos)
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:105]: here also

File: src/mcu/src/HandleFrames.cpp [hunk 6]
```diff
@@ -7,71 +7,450 @@
...
+        }
+        /* Increment sequenceNumber after each concatenation*/
+        expected_sequence_number++;
+         /* Check if all multi-frames have been received */
+        if (frame_data.size() >= expected_payload) 
+        { 
+            std::cout << "data is: ";
+            for (int data_pos = 0; data_pos < (int)frame_data.size(); ++data_pos) 
+            {
+                std::cout << std::hex << int(frame_data[data_pos]) << " ";
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:107]: here

File: src/mcu/src/HandleFrames.cpp [hunk 7]
```diff
@@ -7,71 +7,450 @@
...
 
     switch (sid) {
-    /* UDS Requests */
         case 0x10:
-            /* DiagnosticSessionControl(); */
-            std::cout << "DiagnosticSessionControl called." << std::endl;
+            /* DiagnosticSessionControl(sid, frame_data[2]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:136]: here

File: src/mcu/src/HandleFrames.cpp [hunk 8]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "DiagnosticSessionControl called." << std::endl;
+            }
             break;
         case 0x11:
-            /* EcuReset(); */
+            /* EcuReset(sid, frame_data[2]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:147]: here

File: src/mcu/src/HandleFrames.cpp [hunk 9]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "EcuReset called." << std::endl;
+            }
             break;
         case 0x27:
-            /* SecurityAccess(); */
+            /* SecurityAccess(sid, frame_data[2], key?); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:158]: here

File: src/mcu/src/HandleFrames.cpp [hunk 10]
```diff
@@ -7,71 +7,450 @@
...
+            else 
+            {
+                std::cout << "SecurityAccess called." << std::endl;
+            }
             break;
         case 0x29:
             /* Authentication(); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:169]: here

File: src/mcu/src/HandleFrames.cpp [hunk 11]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "Authentication called." << std::endl;
+            }
             break;
         case 0x3E:
-            /* TesterPresent(); */
+            /* TesterPresent(sid, frame_data[2]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:180]: here

File: src/mcu/src/HandleFrames.cpp [hunk 12]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "TesterPresent called." << std::endl;
+            }
             break;
         case 0x83:
-            /* AccessTimingParameters(); */
+            /* AccessTimingParameters(sid, frame_data[2]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:191]: here

File: src/mcu/src/HandleFrames.cpp [hunk 13]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "AccessTimingParameters called." << std::endl;
+            }
             break;
         case 0x22:
-            /* ReadDataByIdentifier(); */
+            /* ReadDataByIdentifier(sid, frame_data[2] << 8) | frame_data[3]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:202]: hre

File: src/mcu/src/HandleFrames.cpp [hunk 14]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "ReadDataByIdentifier called." << std::endl;
+            }
             break;
         case 0x23:
-            /* ReadMemoryByAddress(); */
+            /* ReadMemoryByAddress(frame_data[2], frame_data[3] << 8) | frame_data[4], frame_data[5] << 8) | frame_data[6]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:213]: here

File: src/mcu/src/HandleFrames.cpp [hunk 15]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "ReadMemoryByAddress called." << std::endl;
+            }
             break;
         case 0x2E:
-            /* WriteDataByIdentifier(); */
+            /* WriteDataByIdentifier(sid, frame_data[2] << 8) | frame_data[3], data_parameter?); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:224]: here

File: src/mcu/src/HandleFrames.cpp [hunk 16]
```diff
@@ -7,71 +7,450 @@
...
+            else 
+            {
+                std::cout << "WriteDataByIdentifier called with one frame." << std::endl;
+            }
             break;
         case 0x14:
             /* ClearDiagnosticInformation(); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:239]: here

File: src/mcu/src/HandleFrames.cpp [hunk 17]
```diff
@@ -7,71 +7,450 @@
...
+            else 
+            {
+                std::cout << "ClearDiagnosticInformation called." << std::endl;
+            }
             break;
         case 0x19:
             /* ReadDtcInformation(); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:250]: here

File: src/mcu/src/HandleFrames.cpp [hunk 18]
```diff
@@ -7,71 +7,450 @@
...
             break;
-
         /* OTA Requests */
         case 0x34:
-            /* RequestDownload(); */
-            std::cout << "Request download called." << std::endl;
+            /* RequestDownload(sid, frame_data[2], frame_data[3], frame_data[4], frame_data[5]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:338]: here

File: src/mcu/src/HandleFrames.cpp [hunk 19]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "RequestDownload called." << std::endl;
+            }
             break;
         case 0x36:
-            /* TransferData(); */
+            /* TransferData(sid, frame_data[2], frame_data[3], frame_data[4]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:349]: here

File: src/mcu/src/HandleFrames.cpp [hunk 20]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "TransferData called with one frame" << std::endl;
+            }
             break;
         case 0x37:
-            /* RequestTransferExit(); */
+            /* RequestTransferExit(sid, frame_data[2]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:364]: here

File: src/mcu/src/HandleFrames.cpp [hunk 21]
```diff
@@ -7,71 +7,450 @@
...
+            else 
+            {
+                std::cout << "RequestTransferExit called." << std::endl;
+            }
             break;
         case 0x32:
             /* RequestUpdateStatus(); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:375]: here

File: src/mcu/src/HandleFrames.cpp [hunk 22]
```diff
@@ -7,71 +7,450 @@
...
\ No newline at end of file
+}
+void HandleFrames::processNrc(canid_t frame_id, uint8_t sid, uint8_t nrc)
+{
+    switch(nrc)
+    {
+        case 0x11:
+            /* Service not supported */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Service not supported for service: " << std::hex << (int)sid << std::endl;
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:412]: here

File: src/mcu/src/HandleFrames.cpp [hunk 23]
```diff
@@ -7,71 +7,450 @@
...
+        break;
+        case 0x13:
+            /* Incorrect message length or invalid format */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Incorrect message length or invalid format for service: " << std::hex << (int)sid << std::endl;
+        break;
+        case 0x14:
+            /*  Response too long */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Response too long for service: " << std::hex << (int)sid << std::endl;
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:422]: here

File: src/mcu/src/HandleFrames.cpp [hunk 24]
```diff
@@ -7,71 +7,450 @@
...
+        break;
+        case 0x25:
+            /* No response from subnet component */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: No response from subnet component for service: " << std::hex << (int)sid << std::endl;
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:427]: here

File: src/mcu/src/HandleFrames.cpp [hunk 25]
```diff
@@ -7,71 +7,450 @@
...
+        break;
+        case 0x34:
+            /* Authentication failed */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Authentication failed for service: " << std::hex << (int)sid << std::endl;
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:432]: here

File: src/mcu/src/HandleFrames.cpp [hunk 26]
```diff
@@ -7,71 +7,450 @@
...
+        break;
+        case 0x94:
+            /* Resource temporarily unavailable */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Resource temporarily unavailable for service: " << std::hex << (int)sid << std::endl;
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:437]: here

File: src/mcu/src/HandleFrames.cpp [hunk 27]
```diff
@@ -7,71 +7,450 @@
...
+        break;
+        case 0x70:
+            /* Upload download not accepted */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Upload download not accepted for service: " << std::hex << (int)sid << std::endl;
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:442]: here

File: src/mcu/src/HandleFrames.cpp [hunk 28]
```diff
@@ -7,71 +7,450 @@
...
+        break;
+        case 0x71:
+            /* Transfer data suspended */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Transfer data suspended for service: " << std::hex << (int)sid << std::endl;
+        break;
+        default:
+            /* Unknown negative response code */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Unknown negative response code for service: " << std::hex << (int)sid << std::endl;
```
Comment (Miruna-Maria) [inline src/mcu/src/HandleFrames.cpp:452]: here

File: src/mcu/src/HandleFrames.cpp (hunk 3 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 5 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 2 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:64]: added for print purposes only, it prints wrong without cast

File: src/mcu/src/HandleFrames.cpp (hunk 4 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:91]: added for print purposes only, it prints wrong without cast

File: src/mcu/src/HandleFrames.cpp (hunk 7 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 9 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 11 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 13 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 15 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 17 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 18 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 20 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 22 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:412]: added for print purposes only, it prints wrong without cast

File: src/mcu/src/HandleFrames.cpp (hunk 23 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:422]: added for print purposes only, it prints wrong without cast

File: src/mcu/src/HandleFrames.cpp (hunk 25 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:432]: added for print purposes only, it prints wrong without cast

File: src/mcu/src/HandleFrames.cpp (hunk 27 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:442]: added for print purposes only, it prints wrong without cast

File: src/mcu/src/HandleFrames.cpp (hunk 28 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:452]: added for print purposes only, it prints wrong without cast

File: src/mcu/src/HandleFrames.cpp (hunk 6 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:107]: done

File: src/mcu/src/HandleFrames.cpp (hunk 10 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 14 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 16 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 19 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp [hunk 29]
```diff
@@ -7,71 +7,450 @@
...
+    {
+        case 0x11:
+            /* Service not supported */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Service not supported for service: " << std::hex << (int)sid << std::endl;
+        break;
+        case 0x13:
+            /* Incorrect message length or invalid format */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Incorrect message length or invalid format for service: " << std::hex << (int)sid << std::endl;
```
[the remark this answers is not in the document]
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:417]: done

File: src/mcu/src/HandleFrames.cpp [hunk 30]
```diff
@@ -7,71 +7,450 @@
...
+        break;
+        case 0x70:
+            /* Upload download not accepted */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Upload download not accepted for service: " << std::hex << (int)sid << std::endl;
+        break;
+        case 0x71:
+            /* Transfer data suspended */
+            //GenerateFrames::negativeResponse(can_id, sid, nrc);
+            std::cout << "Error: Transfer data suspended for service: " << std::hex << (int)sid << std::endl;
```
[the remark this answers is not in the document]
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:447]: done

File: src/mcu/src/HandleFrames.cpp (hunk 26 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:437]: done

File: src/mcu/src/HandleFrames.cpp (hunk 24 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp:427]: done

File: src/mcu/src/HandleFrames.cpp (hunk 21 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp [hunk 31]
```diff
@@ -7,71 +7,450 @@
...
+            {
+                std::cout << "ReadDtcInformation called." << std::endl;
+            }
             break;
         case 0x31:
-            /* RoutineControl(); */
+            /* RoutineControl(sid, frame_data[2], frame_data[3] << 8) | frame_data[4]); */
+            if(frame_data[1] == 0x7F)
+            {
+                processNrc(frame_id, sid, (int)frame_data[3]);
```
[the remark this answers is not in the document]
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 12 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

File: src/mcu/src/HandleFrames.cpp (hunk 8 above)
Comment (Mihnea9901) [reply src/mcu/src/HandleFrames.cpp]: done

**Review by Miruna-Maria — APPROVED** [review]

**Pull request merged**